src/contracts/atlas/GasAccounting.sol

```diff
@@ -201,107 +208,187 @@ abstract contract GasAccounting is SafetyLocks {
+                }
+                claimsPaidToBundler = 0; // Bundler-Solver is in deficit, no need to pay them.
+            } else {
+                // CASE: in surplus, add to bonded balance by crediting the bundler at bottom of function
+                claimsPaidToBundler = _deposits - _withdrawals - netAtlasGasSurcharge;
+            }
+        } else if (_writeoffs + _deposits < _claims + _withdrawals) {
+            // CASE: Not a special bundler, solver successful, balance in deficit.
+            // NOTE _claims and _writeoffs already have the Gas Surcharges factored in
+            uint256 amountOwed = _claims + _withdrawals - _writeoffs - _deposits;
+            uint256 deficit = _assign(_winningSolver, amountOwed, true);
             if (deficit > 0) {
-                revert InsufficientTotalBalance(deficit);
+                // CASE: Solver's bonded balance isn't enough to cover the amount owed, and the
+                // winning solver is unrelated to the bundler. The bundler is not the solver.
+                if (deficit > _claims - _writeoffs - netAtlasGasSurcharge) {
+                    // CASE: The deficit is too large to writeoff by having the bundler absorb the cost
+                    revert InsufficientTotalBalance(deficit);
+                } else {
+                    // CASE: We can writeoff the deficit (bundler has already paid the gas anyway)
+                    // TODO: Ensure incentive compatibility
+                    _writeoffs += deficit;
+                }
             }
+            claimsPaidToBundler = _claims - _writeoffs - netAtlasGasSurcharge;
         } else {
-            // CASE: in surplus, add to bonded balance
-            // TODO: make sure this works w/ the surcharge 10%
-            uint256 amountCredited = _deposits - _claims - _withdrawals;
-            _credit(winningSolver, amountCredited);
+            // CASE: Solver is in surplus, add to bonded balance
+            uint256 amountCredited = _deposits + _writeoffs - _claims - _withdrawals;
+            _credit(_winningSolver, amountCredited);
+            claimsPaidToBundler = _claims - _writeoffs - netAtlasGasSurcharge;
         }
 
-        netGasSurcharge = (_claims * SURCHARGE_RATE) / (SURCHARGE_SCALE + SURCHARGE_RATE);
-
-        _claims -= netGasSurcharge;
-
-        cumulativeSurcharge = _surcharge + netGasSurcharge;
-
-        SafeTransferLib.safeTransferETH(ctx.bundler, _claims);
+        if (claimsPaidToBundler > 0) {
+            if (msg.value == 0) {
+                _credit(ctx.bundler, claimsPaidToBundler);
+            } else if (msg.value < claimsPaidToBundler) {
+                _credit(ctx.bundler, claimsPaidToBundler - msg.value);
+                SafeTransferLib.safeTransferETH(ctx.bundler, msg.value);
+            } else {
+                SafeTransferLib.safeTransferETH(ctx.bundler, claimsPaidToBundler);
+            }
+        }
```

Comment:
```suggestion
        if (claimsPaidToBundler > 0) {
            _credit(ctx.bundler, claimsPaidToBundler);
        }
```
This would resolve all concerns in https://github.com/spearbit-audits/review-fastlane/issues/20 of bundler reverting tx at the end for whatever reason.

Reply:
Above is outdated. We now just have 1 line at the end of `_settle()` which transfers ETH directly to bundler:
`SafeTransferLib.safeTransferETH(ctx.bundler, claimsPaidToBundler);`

While this gives control flow to the bundler, we think this is safe because:

1. `reconcile()` cannot be called at this point
2. The ERC20 functions of AtlETH have been removed, so those are not an attack vector anymore
3. The bundler would still be paying gas for the entire metacall tx, so there is an economic disincentive to not revert.

Furthermore, the UX benefit of sending the refunded ETH directly to the bundler's address are significant over the friction of additional calls to unbond/redeem any credited AtlETH.